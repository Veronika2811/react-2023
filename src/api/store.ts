import { createWrapper } from 'next-redux-wrapper';
import { configureStore } from '@reduxjs/toolkit';

import { apiSlice } from './apiSlice';

export const makeStore = () =>
  configureStore({
    reducer: {
      [apiSlice.reducerPath]: apiSlice.reducer,
    },
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware().concat(apiSlice.middleware),
  });

export type AppStore = ReturnType<typeof makeStore>;
export type RootState = ReturnType<AppStore['getState']>;
export type AppDispatch = AppStore['dispatch'];

export const wrapper = createWrapper<AppStore>(makeStore, { debug: true });
