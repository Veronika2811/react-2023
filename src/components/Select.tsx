import { useContext } from 'react';
import { SetURLSearchParams } from 'react-router-dom';

import { CharactersContext } from '../context/context';

interface ISelectProps {
  setSearchParams: SetURLSearchParams;
}

const Select = ({ setSearchParams }: ISelectProps) => {
  const { setPerPage } = useContext(CharactersContext);

  return (
    <select
      className="select"
      name="count-elements"
      onChange={(e) => {
        setPerPage(+e.target.value);
        setSearchParams((searchParams) => {
          searchParams.set('page', '1');
          return searchParams;
        });
      }}
    >
      <option value="20">20 cards</option>
      <option value="10">10 cards</option>
    </select>
  );
};

export default Select;
