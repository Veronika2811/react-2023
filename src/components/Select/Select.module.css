.select {
  padding: 0.6em 1.2em;
  border: none;
  border-radius: 5px;
}