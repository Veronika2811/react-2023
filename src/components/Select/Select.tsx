import { ChangeEvent } from 'react';
// import { useSearchParams } from 'react-router-dom';

// import { useAppDispatch } from '@/store/hooks';
import { charactersChangePerPage } from '@/store/slice/charactersSlice';
import {
  // DEFAULT_PAGE,
  DEFAULT_VALUE_PER_PAGE,
  // PAGE_URL_PARAMETER_KEY,
} from '@/constants/constants';
import { useRouter } from 'next/router';

import classes from './Select.module.css';

const Select = () => {
  const router = useRouter()

  const { name, perPage, page } = router.query;

  // const [, setSearchParams] = useSearchParams();

  // const dispatch = useAppDispatch();

  const onChangeItemsPerPage = (e: ChangeEvent<HTMLSelectElement>) => {
    const target = e.target.value;

    router.push({
      pathname: '/',
      // query: { name: currentValue || '', page: page || 1 },
      query: { name: name || '', page: 1, perPage: +target || 20 },
      // query: currentValue? { name: currentValue} : {},
    })

    // dispatch(charactersChangePerPage(+target));

    // setSearchParams((searchParams) => {
    //   searchParams.set(PAGE_URL_PARAMETER_KEY, DEFAULT_PAGE);
    //   return searchParams;
    // });
  };

  return (
    <select
      name="count-elements"
      defaultValue={perPage}
      onChange={onChangeItemsPerPage}
      className={classes.select}
      data-testid="select"
    >
      <option value="20" data-testid="option">
        20 cards
      </option>
      <option value="10" data-testid="option">
        10 cards
      </option>
    </select>
  );
};

export default Select;
