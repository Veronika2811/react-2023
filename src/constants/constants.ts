export const BASE_URL = 'https://rickandmortyapi.com/api/character';

export const DEFAULT_VALUE_PER_PAGE = 20;
export const ADDITIONAL_VALUE_PER_PAGE = 10;

export const LOCAL_STORAGE_KEY = 'Veronika2811-react-2023__searchRequest';

export const DETAILS_URL_PARAMETER_KEY = 'details';
export const PAGE_URL_PARAMETER_KEY = 'page';

export const DEFAULT_PAGE = '1';
