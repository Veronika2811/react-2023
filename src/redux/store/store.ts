import {
  PreloadedState,
  combineReducers,
  configureStore,
} from '@reduxjs/toolkit';

import { apiSlice } from '../api/apiSlice';
import reducer from './charactersSlice';

const rootReducer = combineReducers({
  CHARACTERS_SLICE: reducer,
});

export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: {
      CHARACTERS_SLICE: reducer,
      [apiSlice.reducerPath]: apiSlice.reducer,
    },
    preloadedState,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware().concat(apiSlice.middleware),
  });
};

export const store = setupStore();

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
