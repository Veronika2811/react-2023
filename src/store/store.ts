import {
  PreloadedState,
  combineReducers,
  configureStore,
} from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query';

import { apiSlice } from '@/api/apiSlice';
import reducer from './slice/charactersSlice';

const rootReducer = combineReducers({
  CHARACTERS_SLICE: reducer,
});

export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: {
      CHARACTERS_SLICE: reducer,
      [apiSlice.reducerPath]: apiSlice.reducer,
    },
    preloadedState,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware().concat(apiSlice.middleware),
  });
};

export const store = setupStore();

setupListeners(store.dispatch);

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
