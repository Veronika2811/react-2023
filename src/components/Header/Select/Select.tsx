import { useRouter } from 'next/router';
import { ChangeEvent } from 'react';

import { DEFAULT_PAGE } from '@/utils/constants/constants';

import classes from './Select.module.css';

const Select = () => {
  const router = useRouter();

  const { name, perPage } = router.query;

  const onChangeItemsPerPage = (e: ChangeEvent<HTMLSelectElement>) => {
    const target = e.target.value;

    router.push({
      pathname: '/',
      query: {
        ...(name ? { name } : {}),
        page: DEFAULT_PAGE,
        perPage: target,
      },
    });
  };

  return (
    <select
      name="count-elements"
      defaultValue={perPage}
      onChange={onChangeItemsPerPage}
      className={classes.select}
      data-testid="select"
    >
      <option value="20" data-testid="option">
        20 cards
      </option>
      <option value="10" data-testid="option">
        10 cards
      </option>
    </select>
  );
};

export default Select;
