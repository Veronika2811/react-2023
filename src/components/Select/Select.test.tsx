import { expect, it } from 'vitest';
import { HashRouter } from 'react-router-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';

import Select from './Select';
import { store } from '../../redux/store/store';
import { ADDITIONAL_VALUE_PER_PAGE } from '../../constants/constants';

describe('Pagination component', () => {
  it('renders correctly Pagination component', () => {
    const container = render(
      <HashRouter>
        <Provider store={store}>
          <Select />
        </Provider>
      </HashRouter>
    );
    expect(container).toMatchSnapshot();
  });

  it('should check the changes Select component', async () => {
    render(
      <HashRouter>
        <Provider store={store}>
          <Select />
        </Provider>
      </HashRouter>
    );

    const options = screen.getAllByTestId('option') as HTMLOptionElement[];
    expect(options[0].selected).toBeTruthy();
    expect(options[1].selected).toBeFalsy();

    fireEvent.change(screen.getByTestId('select'), {
      target: { value: ADDITIONAL_VALUE_PER_PAGE },
    });
    expect(options[0].selected).toBeFalsy();
    expect(options[1].selected).toBeTruthy();
  });
});
