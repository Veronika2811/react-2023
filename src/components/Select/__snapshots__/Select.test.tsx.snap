// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pagination component > renders correctly Pagination component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <select
        class="_select_e285b3"
        data-testid="select"
        name="count-elements"
      >
        <option
          data-testid="option"
          selected=""
          value="20"
        >
          20 cards
        </option>
        <option
          data-testid="option"
          value="10"
        >
          10 cards
        </option>
      </select>
    </div>
  </body>,
  "container": <div>
    <select
      class="_select_e285b3"
      data-testid="select"
      name="count-elements"
    >
      <option
        data-testid="option"
        selected=""
        value="20"
      >
        20 cards
      </option>
      <option
        data-testid="option"
        value="10"
      >
        10 cards
      </option>
    </select>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
